[Serve] Make `@serve.batch` wait before creating initial batch

shrekris-anyscale opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

`@serve.batch` currently starts tracking `batch_wait_timeout_s` from when the last batch finished. If a request arrives after the queue has been empty for a while (longer than `batch_wait_timeout_s`), the `@serve.batch`-decorated function runs the request immediately. Even if some requests arrive shortly afterwards, they won't be batched with the first request. They must wait for the first request to finish before being run.

With this change, `@serve.batch` tracks `batch_wait_timeout_s` from when the first request for the next batch arrives. Even if a request arrives after the queue has been empty for a while, `@serve.batch` waits `batch_wait_timeout_s` seconds for more requests before running.

## Related issue number

<!-- For example: "Closes #1234" -->

N/A

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - This change adds a unit test to `test_batching.py`.

edoakes commented on python/ray/serve/tests/test_batching.py:
```diff
@@ -214,6 +215,36 @@ async def call(arg):
+    # Run one request, so internal batch queue gets constructed
+    await get_or_create_event_loop().create_task(no_op(None))
+
+    time.sleep(0.1)
```
this needs to await to not block loop

edoakes replied:
also why do we need it -- preferable to just `await` the task created above

shrekris-anyscale replied:
The `time.sleep(0.1)` is necessary to actually test the behavior. Without the fix in this change, the queue needs to be empty for `batch_wait_timeout_s` seconds for `@serve.batch` to just process the first request that shows up.

shrekris-anyscale replied:
> this needs to await to not block loop

What loop is it blocking?

edoakes replied:
The event loop -- the `@serve.batch` code isn't running at all because you didn't `await`

edoakes replied:
> The time.sleep(0.1) is necessary to actually test the behavior. Without the fix in this change, the queue needs to be empty for batch_wait_timeout_s seconds for @serve.batch to just process the first request that shows up.

Can't you just `await` that first request instead of the sleep?

shrekris-anyscale replied:
I think this is resolved by [this comment](https://github.com/ray-project/ray/pull/36510#discussion_r1232842594). Let me know if it's not.

edoakes commented on python/ray/serve/tests/test_batching.py:
```diff
@@ -214,6 +215,36 @@ async def call(arg):
+    done, _ = await asyncio.wait(tasks, timeout=0.01)
+
+    # All the timeout tasks should be finished
+    assert len(done) == 10
+    for task in tasks:
+        assert task in done
+        assert await task == "No-op"
```
maybe better way to write it depending on your preference :)
```suggestion
    assert set(tasks) == set(done)
    assert all(t.result() == "No-op" for t in tasks)
```

shrekris-anyscale replied:
Cool, I didn't know tasks were hashable.

edoakes commented on python/ray/serve/tests/test_batching.py:
```diff
@@ -214,6 +215,36 @@ async def call(arg):
+    async def no_op(requests):
+        return ["No-op"] * len(requests)
+
+    # Run one request, so internal batch queue gets constructed
+    await get_or_create_event_loop().create_task(no_op(None))
```
this is sort of leaking an implementation detail. prefer to just run the full test twice -- that checks that this happens for the first batch and for subsequent batches in the future

shrekris-anyscale replied:
Sounds good, I made the test run twice instead of leaking the implementation detail.

edoakes commented on python/ray/serve/tests/test_batching.py:
```diff
@@ -214,6 +215,36 @@ async def call(arg):
 
+@pytest.mark.asyncio
+async def test_batch_timeout_empty_queue():
+    """Check that Serve waits for before creating the first batch."""
```
specify that it should wait for the timeout _after_ the first request for a new batch arrives rather than after the first batch ends

shrekris-anyscale replied:
Sounds good, I reworded the docstring.

edoakes commented on python/ray/serve/batching.py:
```diff
@@ -97,52 +96,41 @@ def __init__(
+                self.timeout_s - (time.time() - batch_start_time), 0
+            )
+            try:
+                batch.append(
+                    await asyncio.wait_for(self.queue.get(), remaining_batch_time_s)
```
[minor]

you can optimize this a little bit by using `while not self.queue.empty()` and `self.queue.get_nowait()` to avoid unnecessarily yielding when there are results ready immediately

shrekris-anyscale replied:
I added that in the `while` loop afterwards. I felt it was a bit more readable to avoid nested `while` loops here.

edoakes commented on python/ray/serve/batching.py:
```diff
@@ -97,52 +96,41 @@ def __init__(
+
+        # No time left to wait for more queue arrivals. Add any requests
+        # that are already queued until batch is full or queue is empty.
+        while len(batch) < self.max_batch_size and not self.queue.empty():
+            batch.append(self.queue.get_nowait())
```
oh, I think you can just put this in the main loop body.

make the `wait_for` call `queue.wait()` instead, then after the `wait_for`, just always run this loop

shrekris-anyscale replied:
That would be ideal, but `asyncio.Queue` doesn't have a `wait` function. The two `while` loops is my workaround.

edoakes commented on python/ray/serve/tests/test_batching.py:
```diff
@@ -214,6 +215,39 @@ async def call(arg):
+        assert all(t.result() == "No-op" for t in tasks)
+
+        if iteration < num_iterations - 1:
+            # Leave queue empty for batch_wait_timeout_s between batches
```
oh I get it now -- without the change, this would immediately return the first request enqueued

shrekris-anyscale replied:
Yeah, exactly.

shrekris-anyscale commented:
Synced with @edoakes offline. I moved the second `while` loop inside the first.

I also introduced an `asyncio.Event()` that's set whenever requests are put in the queue. I'm calling `asyncio.wait_for()` on the event instead of on `self.queue.get(...)` directly, since it's unclear whether `asyncio.wait_for(queue.get(), timeout)` introduces a race condition.

shrekris-anyscale commented:
The test failures are unrelated:

* Windows Build & Test ([`test_client_library_integration`](https://buildkite.com/ray-project/oss-ci-build-pr/builds/25986#0188d757-8e9b-46f7-861a-cd56eee95c7f/7759-8678)) was failing on `master`:

<img width="1378" alt="Screen Shot 2023-06-20 at 9 11 48 AM" src="https://github.com/ray-project/ray/assets/92341594/db061fd1-c162-46c9-a41a-58f29d01a5ef">

* [`Documentation`](https://buildkite.com/ray-project/oss-ci-build-pr/builds/25986#0188d760-b00f-44f9-9836-a07434a44c5b/1402-2439) is failing on `master`:

<img width="1330" alt="Screen Shot 2023-06-20 at 9 12 17 AM" src="https://github.com/ray-project/ray/assets/92341594/39bc06f6-885f-47bf-882d-bc35a4cc2bf1">

shrekris-anyscale commented:
@edoakes @architkulkarni This change is ready to merge.

architkulkarni commented:
> @edoakes @architkulkarni This change is ready to merge.

Looks like @edoakes needs to approve it first.
